## datetime module





The `datetime` module provides six main high-level interface classes:

- `date`: An idealized date that assumes the Gregorian calendar extends infinitely into the future and past. It stores the year, month, and day as attributes.
- `time`: An idealized time that assumes there are $24 \times 60 \times 60=86,400$ seconds per day with no leap seconds. This object stores the hour, minute, second, microsecond, and tzinfo (time zone information).
- `datetime`: A combination of a date and a time. It has all the attributes of both classes.
- `timedelta`: A duration expressing the difference between two date, time, or datetime objects to microsecond resolution.
- `tzinfo`: Provides timezone-related information objects.
- `timezone`: A class that implements the `tzinfo` abstract base class as a fixed offset from the UTC.

There is also a `calendar` module for dealing with calendar functions
- month and week arithmetic
- formatting outputs

In [1]:
import datetime

**What is inside `datetime`?**

In [2]:
print(dir(datetime))

['MAXYEAR', 'MINYEAR', 'UTC', '__all__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'date', 'datetime', 'datetime_CAPI', 'time', 'timedelta', 'timezone', 'tzinfo']


## Times

- Time values are represented with the **time** class. 
- Times have attributes for hour, minute, second, and microsecond. 

```python
datetime.time(hour=0, minute=0, second=0, microsecond=0, 
              tzinfo=None, *, fold=0)
```
more information in the [documentation](https://docs.python.org/3/library/time.html)

In [3]:
t=datetime.time(1, 2, 3)

In [4]:
print(t)

01:02:03


In [5]:
print (f'hour  :      {t.hour}')
print (f'minute:      {t.minute}')
print (f'second:      {t.second}')
print (f'microsecond: {t.microsecond}')
print (f'tzinfo:      {t.tzinfo}')

hour  :      1
minute:      2
second:      3
microsecond: 0
tzinfo:      None


The variable <B>t</B> only holds values of time, and not a date associated with the time.
<P>
You can get the valid range of times in a single day:

In [6]:
print ('Earliest  :', datetime.time.min)
print ('Latest    :', datetime.time.max)
print ('Resolution:', datetime.time.resolution)

Earliest  : 00:00:00
Latest    : 23:59:59.999999
Resolution: 0:00:00.000001


Note that the resolution for time is limited to whole microseconds.

---

### <font color="blue">Breakout 1</font>
Create a `time` object with 5 microseconds and 39 minutes.

<p>

<details><summary><b>Click here to access the solution</b></summary>
<p>


```python
t = datetime.time(microsecond=5, minute=29)
print(t)
```

</p>
</details>

---

In [41]:
t = datetime.time(microsecond=5, minute=29)
print(t)

00:29:00.000005


## Dates

- Calendar date values are represented with the **date** class.
- A date is represented in terms of year, month, and day. 
- It is easy to create a date representing today’s date using the **today()** class method.
    
```python
datetime.date(year, month, day)
```
More information in the [documentation](https://docs.python.org/3/library/datetime.html#date-objects)

In [7]:
today = datetime.date.today()
print (f'today: \n\t {today}')
print (f'ctime: \n\t {today.ctime()}')
print (f'tuple: \n\t {today.timetuple()}')
print (f'ordinal: \n\t {today.toordinal()}')
print (f'Year: \n\t {today.year}')
print (f'Mon: \n\t {today.month}')
print (f'Day: \n\t {today.day}')

today: 
	 2024-11-04
ctime: 
	 Mon Nov  4 00:00:00 2024
tuple: 
	 time.struct_time(tm_year=2024, tm_mon=11, tm_mday=4, tm_hour=0, tm_min=0, tm_sec=0, tm_wday=0, tm_yday=309, tm_isdst=-1)
ordinal: 
	 739194
Year: 
	 2024
Mon: 
	 11
Day: 
	 4


There are also class methods for creating instances from integers (using proleptic Gregorian ordinal values, which starts counting from Jan. 1 of the year 1) or POSIX timestamp values.

The following example illustrates the different value types used by:

- `fromordinal()`: Return the date corresponding to the proleptic Gregorian ordinal, (number of days elapsed from the date 01/Jan/0001)
- `fromtimestamp()`: Return the local date corresponding to the timestamp.

In [8]:
import time

o = 733114
print (f'o = {o} and fromordinal(o) = {datetime.date.fromordinal(o)}')

t = time.time()
print (f't = {t} and fromtimestamp(t) = {datetime.date.fromtimestamp(t)}')

o = 733114 and fromordinal(o) = 2008-03-13
t = 1730701232.7725916 and fromtimestamp(t) = 2024-11-04


We can also determine the range of date values:

In [9]:
print (f'Earliest:   {datetime.date.min}')
print (f'Latest:     {datetime.date.max}')
print (f'Resolution: {datetime.date.resolution}')

Earliest:   0001-01-01
Latest:     9999-12-31
Resolution: 1 day, 0:00:00


Note too that the resolution for dates is a whole day.
<P>
Another way to create new date instances uses the <B>replace()</B> method of an existing date. For example, you can change the year, leaving the day and month alone.

In [10]:
d1 = datetime.date(2022, 7, 5)
print ('d1:', d1)

d2 = d1.replace(year=2023)
print ('d2:', d2)

d1: 2022-07-05
d2: 2023-07-05


In [11]:
help(d1.replace)

Help on built-in function replace:

replace(...) method of datetime.date instance
    Return date with new specified fields.



---

### <font color="blue">Breakout 2</font>
Create a `date` object where year is 2023 and the day is 5.

<p>

<details><summary><b>Click here to access the solution</b></summary>
<p>


```python
d = datetime.time(day=5, year=2023)
print(d)
```

</p>
</details>

---

In [43]:
d = datetime.date(day=5, month=9, year=2023)
print(d)

2023-09-05


## timedeltas

- Timedelta objects represent a duration.
- Can be used to calculate the difference between two dates or times.
- We can use <B>datetime</B> to perform basic arithmetic on date values via the <B>timedelta</B> class.

```python
datetime.timedelta(days=0, seconds=0, microseconds=0, 
                   milliseconds=0, minutes=0, hours=0, weeks=0)
```

In [12]:
print (f"microseconds: {datetime.timedelta(microseconds=1)}")
print (f"milliseconds: {datetime.timedelta(milliseconds=1)}")
print (f"seconds:      {datetime.timedelta(seconds=1)}")
print (f"minutes:      {datetime.timedelta(minutes=1)}")
print (f"hours:        {datetime.timedelta(hours=1)}")
print (f"days:         {datetime.timedelta(days=1)}")
print (f"weeks:        {datetime.timedelta(weeks=1)}")

microseconds: 0:00:00.000001
milliseconds: 0:00:00.001000
seconds:      0:00:01
minutes:      0:01:00
hours:        1:00:00
days:         1 day, 0:00:00
weeks:        7 days, 0:00:00


### <font color='green'>Date Arithmetic</font>

- Date math uses the standard arithmetic operators. 
- The following example with date objects illustrates using <B>timedelta</B> objects to compute new dates, and subtracting date instances to produce timedeltas (including a negative delta value).


In [13]:
today = datetime.date.today()
print('Today:', today)

Today: 2024-11-04


In [14]:
one_timedelta_day = datetime.timedelta(days=1)
print('One timedelta day:', one_timedelta_day)

One timedelta day: 1 day, 0:00:00


In [15]:
yesterday = today - one_timedelta_day
print('Yesterday:', yesterday)

Yesterday: 2024-11-03


In [16]:
tomorrow = today + one_timedelta_day
print('Tomorrow:', tomorrow)

Tomorrow: 2024-11-05


In [17]:
print('tomorrow - yesterday:', tomorrow - yesterday)
print('yesterday - tomorrow:', yesterday - tomorrow)

tomorrow - yesterday: 2 days, 0:00:00
yesterday - tomorrow: -2 days, 0:00:00


---

### <font color="blue">Breakout 3</font>
Create an arbitrary `date` object and add five weeks. Print the new date.

<p>

<details><summary><b>Click here to access the solution</b></summary>
<p>


```python
date = datetime.date.today()
print(date)
new_date = date + datetime.timedelta(weeks=5)
print(new_date)
```

</p>
</details>

---

In [44]:
date = datetime.date.today()
print(date)
new_date = date + datetime.timedelta(weeks=5)
print(new_date)

2024-11-04
2024-12-09


## Comparing Values

Both date and time values can be compared using the standard operators to determine which is earlier or later.

In [18]:
print ('Times:')
t1 = datetime.time(12, 55, 0)
print ('\tt1:', t1)
t2 = datetime.time(13, 5, 0)
print ('\tt2:', t2)
print ('\tt1 < t2:', t1 < t2)

Times:
	t1: 12:55:00
	t2: 13:05:00
	t1 < t2: True


In [19]:
print ('Dates:')
d1 = datetime.date.today()
print ('\td1:', d1)
d2 = datetime.date.today() + datetime.timedelta(days=1)
print ('\td2:', d2)
print ('\td1 > d2:', d1 > d2)

Dates:
	d1: 2024-11-04
	d2: 2024-11-05
	d1 > d2: False


## Combining Dates and Times:  `datetime`  Class

- We can use the datetime class to hold values consisting of both date and time components. 
- Represents all the attributes from date and time objects.

```python
datetime.datetime(year, month, day, 
                  hour=0, minute=0, second=0, 
                  microsecond=0, tzinfo=None, *, fold=0)
```
more infomation in the [documentation](https://docs.python.org/3/library/datetime.html)

In [20]:
print(f'Now:     {datetime.datetime.now()}')
print(f'Today:   {datetime.datetime.today()}')
print(f'UTC Now: {datetime.datetime.utcnow()}')

Now:     2024-11-04 06:20:33.011662
Today:   2024-11-04 06:20:33.011662
UTC Now: 2024-11-04 06:20:33.011662


C:\Users\curleys\AppData\Local\Temp\ipykernel_17700\2921065448.py:3: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  print(f'UTC Now: {datetime.datetime.utcnow()}')


In [21]:
d = datetime.datetime.now()
for attr in [ 'year', 'month', 'day', 'hour', 'minute', 'second', 'microsecond']:
    print(f"{attr}: {getattr(d, attr)}")

year: 2024
month: 11
day: 4
hour: 6
minute: 20
second: 33
microsecond: 24052


In [22]:
NOW = datetime.datetime.now()
 
print(f"Current date & time =         {NOW}")
print(f"Date and time in ISO format = {NOW.isoformat()}")
print(f"Current year =                {NOW.year}")
print(f"Current month =               {NOW.month}")
print(f"Current date (day) =          {NOW.day}")
print(f"dd/mm/yyyy format =           {NOW.day}/{NOW.month}/{NOW.year}")
print(f"Current hour =                {NOW.hour}")
print(f"Current minute =              {NOW.minute}")
print(f"Current second =              {NOW.second}")
print(f"hh:mm:ss format =             {NOW.hour}:{NOW.month}:{NOW.second}")

Current date & time =         2024-11-04 06:20:33.041528
Date and time in ISO format = 2024-11-04T06:20:33.041528
Current year =                2024
Current month =               11
Current date (day) =          4
dd/mm/yyyy format =           4/11/2024
Current hour =                6
Current minute =              20
Current second =              33
hh:mm:ss format =             6:11:33


* Just as with date, datetime provides convenient class methods for creating new instances. It also includes `fromordinal()` and `fromtimestamp()`. 
* In addition, `combine()` can be useful if you already have a date instance and time instance and want to create a datetime.

In [23]:
t = datetime.time(1, 2, 3)
print ('t :', t)

d = datetime.date.today()
print ('d :', d)

dt = datetime.datetime.combine(d, t)
print ('dt:', dt)

t : 01:02:03
d : 2024-11-04
dt: 2024-11-04 01:02:03


## Formating and Parsing

* The default string representation of a datetime object uses the format: `YYYY-MM-DDTHH:MM:SS.mmmmmm` 
* Alternate formats can be generated using `strftime()` function. 
     - Converts datetime objects into strings.
     - Uses different control code to give an output.
     - Each control code resembles different parameters like year, month, weekday and date.
* If your input data includes timestamp values parsable with `time.strptime()`, then `datetime.strptime()` is a convenient way to convert them to datetime instances.

**Useful `strptime` and `strftime` Patterns**

|Directive | Meaning |
| --- | --- |
| `%a` | Weekday as locale's abbreviated name |
| `%A` | Weekday as locale's full name |
| `%w` | Weekday as decimal number, where 0 is Sunday and 6 is Saturday |
| `%d` | Day of the month as a zero-padded decimal number [01, 02, ..., 30, 31] |
| `%b` | Month as locale's abbreviated name |
| `%B` | Weekday as locale's full name |
| `%m` | Month as zero-padded decimal number [01, 02, ..., 11, 12] |
| `%y` | Year without century as a zero-padded decimal number name [0001, ..., 2019, 2020, ..., 9999] |
| `%Y` | Year with century as a decimal number  |
| `%H` | Hour (24-hour clock) as a zero-padded decimal number [00, 01, ..., 22, 23] |
| `%I` | Hour (12-hour clock) as a zero-padded decimal number |
| `%p` | Locale equivalent of either AM or PM |
| `%M` | Minute as a zero-padded decimal number [00, 01, ..., 58, 59]|
| `%S` | Second as a zero-padded decimal number [00, 01, ..., 58, 59]|
| `%f` | Microsecond as a zero-padded decimal number |
| `%j` | Day of the year as a zero-padded decimal number |
| `%W` | Week number of the year (Monday as the first day of the week) as a decimal number |
| `%U` | Week number of the year (Sunday as the first day of the week) as a decimal number |
| `%c` | Locale’s appropriate date and time representation |
| `%Z` | Time zone name |
| `%z` | UTC offset in the form HH[SS[.fffff]] |


**Formatting**

Weekday Month Day Hour:Minute:Second Year

In [24]:
format = "%a %d %b %H:%M:%S %Y"

today = datetime.datetime.today()
print (f'ISO: {today}')

s = today.strftime(format)
print (f'strftime: {s}')

ISO: 2024-11-04 06:20:33.073931
strftime: Mon 04 Nov 06:20:33 2024


Obtain the time in HH:MM:SS

In [25]:
print(today.strftime("%X"))

06:20:33


Obtain the hour with 12 hours time

In [26]:
print(today.strftime("%I"))

06


Obtain AM or PM

In [27]:
print(today.strftime("%p"))

AM


`%c` - local date and time, `%x`-local's date, `%X`- local's time

In [28]:
print(f'Date and Time = {today.strftime("%c")}')
print(f'Date =          {today.strftime("%x")}')
print(f'Time =          {today.strftime("%X")}')

Date and Time = Mon Nov  4 06:20:33 2024
Date =          11/04/24
Time =          06:20:33


`%I/%H` - 12/24 Hour, `%M` - minute, `%S` - second, `%p` - local's AM/PM

In [29]:
print(f"Time = {today.strftime('%I:%M:%S %p')}") # 12-Hour:Minute:Second:AM
print(f"Hour:Minutes = {today.strftime('%H:%M')}") # 24-Hour:Minute

Time = 06:20:33 AM
Hour:Minutes = 06:20


**Parsing**

If the format of a string is known, it can be easily parsed to a `datetime` object using `strptime`.

In [30]:
datetime_str = 'Sep 19 2020 2:42PM'
datetime_fmt = '%b %d %Y %I:%M%p'
datetime_obj = datetime.datetime.strptime(datetime_str, datetime_fmt)

print('Date:        ', datetime_obj.date())
print('Time:        ', datetime_obj.time())
print('Date-time:   ', datetime_obj)

Date:         2020-09-19
Time:         14:42:00
Date-time:    2020-09-19 14:42:00


In [31]:
print('String date: ', datetime_obj.strftime(datetime_fmt))

String date:  Sep 19 2020 02:42PM


Examples of commonly used time formats and the tokens used for parsing:

```python
"Sep 15 2020 at 2:42PM"        -> "%b %d %Y at %I:%M%p"
"September 15, 2020, 14:42:15" -> "%B %d, %Y, %H:%M:%S"
"Tue,09/15/20,2:42PM"          -> "%a,%d/%m/%y,%I:%M%p"
"Tue, 15 September, 2020"      -> "%a, %d %B, %Y"
"2020-09-15T14:42:15Z"         -> "%Y-%m-%dT%H:%M:%SZ"
```

---

### <font color="blue">Breakout 4</font>

Write a function (named `number_days_between`) that:

- Takes two arguments that are 8-digit integers of the form YYYYMMDD (actually a date), and
- Returns the number of days between the two dates.

For instance:

```python
   number_days_between(20200617, 20200619) = 2
   number_days_between(20200619, 20100219) = 3773
```

<p>

<details><summary><b>Click here to access the solution</b></summary>
<p>


```python
def number_days_between(date1, date2):
    """
    Determine the number of days between two dates
    represented as 8-digit integers (YYYYMMDD).
    
    Paramters
    ---------
date = datetime.date.today()
print(date)
new_date = date + datetime.timedelta(weeks=5)
print(new_date)
```

</p>
</details>
    
---

In [45]:
date = datetime.date.today()
print(date)
new_date = date + datetime.timedelta(weeks=5)
print(new_date)

2024-11-04
2024-12-09


---

### <font color="blue">Breakout 5</font>

On July 16, 1969, the huge, 363-feet tall Saturn V rocket launches on the Apollo 11 mission from Pad A, Launch Complex 39, Kennedy Space Center, at 9:32 a.m. EDT. Write a Python program the computes (**from now**) the number of:

- Years
- Months
- Days
- Hours
- Minutes
- Seconds

since the launch.

Hint: [Check this website](https://stackoverflow.com/questions/1345827/how-do-i-find-the-time-difference-between-two-datetime-objects-in-python)
<p>

<details><summary><b>Click here to access the solution</b></summary>
<p>


```python
current_date = datetime.datetime.now()
launch_date = datetime.datetime(1969, 7, 16, 9, 32)
duration = current_date - launch_date

duration_in_s = duration.total_seconds()
years = divmod(duration_in_s, 31536000)[0]  # Seconds in a year=365*24*60*60 = 31536000
days  = divmod(duration_in_s, 86400)[0]     # Seconds in a day = 86400
hours = divmod(duration_in_s, 3600)[0]      # Seconds in an hour = 3600
minutes = divmod(duration_in_s, 60)[0]      # Seconds in a minute = 60

print(f"Number of Years:   {years}")
print(f"Number of Weeks:   {days/7}")
print(f"Number of Days:    {days}")
print(f"Number of Hours:   {hours}")
print(f"Number of Minutes: {minutes}")
print(f"Number of Seconds: {duration_in_s}")
```

</p>
</details>
    
---

In [46]:
current_date = datetime.datetime.now()
launch_date = datetime.datetime(1969, 7, 16, 9, 32)
duration = current_date - launch_date

duration_in_s = duration.total_seconds()
years = divmod(duration_in_s, 31536000)[0]  # Seconds in a year=365*24*60*60 = 31536000
days  = divmod(duration_in_s, 86400)[0]     # Seconds in a day = 86400
hours = divmod(duration_in_s, 3600)[0]      # Seconds in an hour = 3600
minutes = divmod(duration_in_s, 60)[0]      # Seconds in a minute = 60

print(f"Number of Years:   {years}")
print(f"Number of Weeks:   {days/7}")
print(f"Number of Days:    {days}")
print(f"Number of Hours:   {hours}")
print(f"Number of Minutes: {minutes}")
print(f"Number of Seconds: {duration_in_s}")

Number of Years:   55.0
Number of Weeks:   2885.5714285714284
Number of Days:    20199.0
Number of Hours:   484796.0
Number of Minutes: 29087813.0
Number of Seconds: 1745268831.35278


# Here endeth my lecture 
but if you want more

## <font color="red">Bonus Breakout</font>

- We want to write a function that increments a date by adding a number of years, a number of months and a number of days.
- We break the task into small ones by writing three functions that add the days, the months and the years, respectively. We then use the functions to accomplish our task.

### <font color="blue">Problem 1</font>

Write a function (`add_years`) that adds `num_years` to a date object.

<details><summary><b>Click here to access the solution</b></summary>
<p>


```python
def add_years (my_date, num_years):
    return my_date.replace(year=my_date.year + num_years)
```

</p>
</details>

### <font color="blue">Problem 2</font>

Write a function (`add_months`) that adds `num_months` to a date object.

<details><summary><b>Click here to access the solution</b></summary>
<p>


```python
def add_Months(my_date, num_months):
    m = my_date.month + num_months
    year  = m // 12
    month = m % 12
    if month == 0:
       month = 12
       year -= 1
    try:
        new_date = datetime.date(my_date.year+year, month, my_date.day)
    except ValueError:
        import calendar
        # determine the total number of days in a new month
        m1, d1 = calendar.monthrange(my_date.year+year, month)
        new_date = datetime.date(my_date.year+year, month, d1)
    return new_date
```

</p>
</details>

### <font color="blue">Problem 3</font>

Write a function (`add_days`) that adds `num_days` to a date object.

<details><summary><b>Click here to access the solution</b></summary>
<p>


```python
def add_days (my_date, num_days):
    return my_date + datetime.timedelta(days=num_days)
```

</p>
</details>

### <font color="blue">Problem 4</font>

Write a function (`increment_date`) that:

- Has as arguments `ref_date` (in the format YYYYMMDD), `num_years`, `num_months` and `num_days`. 
- Adds `num_years`, `num_months` and `num_days` to `ref_date`.
- Returns a new date in the format YYYYMMDD.

Note that `num_years`, `num_months` and `num_days` can be either positive or negative numbers.

<details><summary><b>Click here to access the solution</b></summary>
<p>


```python
def increment_date(ref_date, num_years=0, num_months=0, num_days=0):
    # Extract the year, the month and day from ref_date
    y = ref_date // 10000
    m = (ref_date % 10000) // 100
    d = ref_date % 100
    
    # Determine the current date object
    cur_date = datetime.date(y, m, d)
    
    # Increment the date
    cur_date = add_years(cur_date, num_years)
    cur_date = add_months(cur_date, num_months)
    cur_date = add_days(cur_date, num_days)
    
    # Extract the new year, month and day
    curY = cur_date.year
    curM = cur_date.month
    curD = cur_date.day
    
    # Compute the new date in the format YYYYMMDD
    new_date = 10000*curY + 100*curM + curD
    
    return new_date

print(increment_date(20001231, num_months=-15, num_days=7))
```

</p>
</details>

## <font color="purple">Going Deeper: Handling timezone</font>

- We can use the `pytz` module that allows accurate and cross platform timezone calculations.
- It provides implementations of **tzinfo** class that can be used to create timezone aware datetime instances.

In [32]:
import pytz
print(dir(pytz))

['AmbiguousTimeError', 'BaseTzInfo', 'FixedOffset', 'HOUR', 'InvalidTimeError', 'LazyDict', 'LazyList', 'LazySet', 'NonExistentTimeError', 'OLSEN_VERSION', 'OLSON_VERSION', 'UTC', 'UnknownTimeZoneError', 'VERSION', 'ZERO', '_CountryNameDict', '_CountryTimezoneDict', '_FixedOffset', '_UTC', '__all__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '__version__', '_all_timezones_lower_to_standard', '_all_timezones_unchecked', '_case_insensitive_zone_lookup', '_p', '_test', '_tzinfo_cache', '_unmunge_zone', 'all_timezones', 'all_timezones_set', 'ascii', 'build_tzinfo', 'common_timezones', 'common_timezones_set', 'country_names', 'country_timezones', 'datetime', 'exceptions', 'lazy', 'open_resource', 'os', 'resource_exists', 'sys', 'timezone', 'tzfile', 'tzinfo', 'unicode', 'unpickler', 'utc']


Commonly Used Time Zones:

In [33]:
print(pytz.common_timezones_set)

LazySet({'Asia/Ashgabat', 'America/Argentina/Mendoza', 'America/Metlakatla', 'Asia/Khandyga', 'America/Campo_Grande', 'America/Guyana', 'Pacific/Gambier', 'Africa/Khartoum', 'Europe/Ulyanovsk', 'Africa/Malabo', 'America/Recife', 'America/Cayenne', 'Europe/Minsk', 'America/Monterrey', 'Atlantic/St_Helena', 'Africa/Maseru', 'Atlantic/Canary', 'Europe/Isle_of_Man', 'America/St_Vincent', 'Asia/Bangkok', 'America/Rankin_Inlet', 'Africa/Bissau', 'Australia/Perth', 'America/Sitka', 'Asia/Sakhalin', 'Europe/Zagreb', 'Asia/Ust-Nera', 'Europe/Dublin', 'US/Eastern', 'Europe/Belgrade', 'Europe/Guernsey', 'Africa/El_Aaiun', 'America/Dawson', 'Antarctica/Macquarie', 'Asia/Kabul', 'Africa/Nairobi', 'America/Menominee', 'Europe/Helsinki', 'Africa/Kampala', 'Africa/Nouakchott', 'Canada/Eastern', 'Antarctica/Vostok', 'Asia/Dubai', 'Pacific/Guadalcanal', 'Africa/Bamako', 'Asia/Qostanay', 'Europe/Malta', 'Africa/Maputo', 'Pacific/Majuro', 'America/Bahia_Banderas', 'America/Nassau', 'Asia/Qatar', 'Asia/Yak

Country Names:

In [34]:
for key, val in pytz.country_names.items():
    print(key, '=', val, end=', ')

AD = Andorra, AE = United Arab Emirates, AF = Afghanistan, AG = Antigua & Barbuda, AI = Anguilla, AL = Albania, AM = Armenia, AO = Angola, AQ = Antarctica, AR = Argentina, AS = Samoa (American), AT = Austria, AU = Australia, AW = Aruba, AX = Åland Islands, AZ = Azerbaijan, BA = Bosnia & Herzegovina, BB = Barbados, BD = Bangladesh, BE = Belgium, BF = Burkina Faso, BG = Bulgaria, BH = Bahrain, BI = Burundi, BJ = Benin, BL = St Barthelemy, BM = Bermuda, BN = Brunei, BO = Bolivia, BQ = Caribbean NL, BR = Brazil, BS = Bahamas, BT = Bhutan, BV = Bouvet Island, BW = Botswana, BY = Belarus, BZ = Belize, CA = Canada, CC = Cocos (Keeling) Islands, CD = Congo (Dem. Rep.), CF = Central African Rep., CG = Congo (Rep.), CH = Switzerland, CI = Côte d'Ivoire, CK = Cook Islands, CL = Chile, CM = Cameroon, CN = China, CO = Colombia, CR = Costa Rica, CU = Cuba, CV = Cape Verde, CW = Curaçao, CX = Christmas Island, CY = Cyprus, CZ = Czech Republic, DE = Germany, DJ = Djibouti, DK = Denmark, DM = Dominica,

Time zones in each country:

In [35]:
print('country_timezones =')
for key, val in pytz.country_timezones.items():
    print(key, '=', val, end=', ')

country_timezones =
AD = ['Europe/Andorra'], AE = ['Asia/Dubai'], AF = ['Asia/Kabul'], AG = ['America/Antigua'], AI = ['America/Anguilla'], AL = ['Europe/Tirane'], AM = ['Asia/Yerevan'], AO = ['Africa/Luanda'], AQ = ['Antarctica/McMurdo', 'Antarctica/Casey', 'Antarctica/Davis', 'Antarctica/DumontDUrville', 'Antarctica/Mawson', 'Antarctica/Palmer', 'Antarctica/Rothera', 'Antarctica/Syowa', 'Antarctica/Troll', 'Antarctica/Vostok'], AR = ['America/Argentina/Buenos_Aires', 'America/Argentina/Cordoba', 'America/Argentina/Salta', 'America/Argentina/Jujuy', 'America/Argentina/Tucuman', 'America/Argentina/Catamarca', 'America/Argentina/La_Rioja', 'America/Argentina/San_Juan', 'America/Argentina/Mendoza', 'America/Argentina/San_Luis', 'America/Argentina/Rio_Gallegos', 'America/Argentina/Ushuaia'], AS = ['Pacific/Pago_Pago'], AT = ['Europe/Vienna'], AU = ['Australia/Lord_Howe', 'Antarctica/Macquarie', 'Australia/Hobart', 'Australia/Melbourne', 'Australia/Sydney', 'Australia/Broken_Hill', 'Austra

Time zones in the US:

In [36]:
print('Supported timezones by US =', pytz.country_timezones['US'])

Supported timezones by US = ['America/New_York', 'America/Detroit', 'America/Kentucky/Louisville', 'America/Kentucky/Monticello', 'America/Indiana/Indianapolis', 'America/Indiana/Vincennes', 'America/Indiana/Winamac', 'America/Indiana/Marengo', 'America/Indiana/Petersburg', 'America/Indiana/Vevay', 'America/Chicago', 'America/Indiana/Tell_City', 'America/Indiana/Knox', 'America/Menominee', 'America/North_Dakota/Center', 'America/North_Dakota/New_Salem', 'America/North_Dakota/Beulah', 'America/Denver', 'America/Boise', 'America/Phoenix', 'America/Los_Angeles', 'America/Anchorage', 'America/Juneau', 'America/Sitka', 'America/Metlakatla', 'America/Yakutat', 'America/Nome', 'America/Adak', 'Pacific/Honolulu']


Get the current local date/time:

In [37]:
local = datetime.datetime.now()
print("Local:", local.strftime("%m/%d/%Y, %H:%M:%S"))
print("Time zone info:", local.tzinfo)

Local: 11/04/2024, 06:20:33
Time zone info: None


Get the current UTC date/time:

In [38]:
datetime_UTC = datetime.datetime.now(pytz.utc)
print("UTC:", datetime_UTC.strftime("%m/%d/%Y, %H:%M:%S"))
print("Time zone info:", datetime_UTC.tzinfo)

UTC: 11/04/2024, 06:20:33
Time zone info: UTC


Get the current date/time in the `America/Bahia_Banderas` time zone:

In [39]:
tz_BB = pytz.timezone('America/Bahia_Banderas') 
datetime_BB = datetime.datetime.now(tz_BB)
print("Bahia Banderas:", datetime_BB.strftime("%m/%d/%Y, %H:%M:%S"))
print("Time zone info:", datetime_BB.tzinfo)

Bahia Banderas: 11/04/2024, 00:20:33
Time zone info: America/Bahia_Banderas


Get the current date/time in the `Europe/London` time zone:

In [40]:
tz_London = pytz.timezone('Europe/London')
datetime_London = datetime.datetime.now(tz_London)
print("London:", datetime_London.strftime("%m/%d/%Y, %H:%M:%S"))
print("Time zone info:", datetime_London.tzinfo)

London: 11/04/2024, 06:20:33
Time zone info: Europe/London


## Reference Documents

- <A HREF="http://pleac.sourceforge.net/pleac_python/datesandtimes.html">Dates and Times</A>
- <A HREF="http://www.marinamele.com/2014/03/13-useful-tips-about-python-datetime.html">Usedul Tips about Python datetime Objects</A>
- <A HREF="https://pymotw.com/3/datetime/">datetime - Date and Time Value Manipulation</A>
- [google colab](https://colab.research.google.com/github/astg606/py_materials/blob/master/datetime/introduction_datetime.ipynb)